src/scene/gsplat/gsplat-data.js

```diff
@@ -12,70 +12,173 @@ const quat2 = new Quat();
- * @property {number} rz - The z-component of the quaternion rotation.
- * @property {number} rw - The w-component of the quaternion rotation.
- * @property {number} sx - The scale factor in the x-direction.
- * @property {number} sy - The scale factor in the y-direction.
- * @property {number} sz - The scale factor in the z-direction.
- */
+// iterator for accessing compressed splat data
+class SplatCompressedIterator {
+    constructor(gsplatData, p, r, s, c) {
+        const unpackUnorm = (value, bits) => {
+            const t = (1 << bits) - 1;
+            return (value & t) / t;
+        };
+
+        const unpack111011 = (result, value) => {
+            result.x = unpackUnorm(value >>> 21, 11);
+            result.y = unpackUnorm(value >>> 11, 10);
+            result.z = unpackUnorm(value, 11);
+        };
+
+        const unpack8888 = (result, value) => {
+            result.x = unpackUnorm(value >>> 24, 8);
+            result.y = unpackUnorm(value >>> 16, 8);
+            result.z = unpackUnorm(value >>> 8, 8);
+            result.w = unpackUnorm(value, 8);
+        };
+
+        // unpack quaternion with 2,10,10,10 format (largest element, 3x10bit element)
+        const unpackRot = (result, value) => {
+            const norm = 1.0 / (Math.sqrt(2) * 0.5);
+            const a = (unpackUnorm(value >>> 20, 10) - 0.5) * norm;
+            const b = (unpackUnorm(value >>> 10, 10) - 0.5) * norm;
+            const c = (unpackUnorm(value, 10) - 0.5) * norm;
+            const m = Math.sqrt(1.0 - (a * a + b * b + c * c));
+
+            switch (value >>> 30) {
+                case 0: result.set(m, a, b, c); break;
+                case 1: result.set(a, m, b, c); break;
+                case 2: result.set(a, b, m, c); break;
+                case 3: result.set(a, b, c, m); break;
+            }
+        };
+
+        const lerp = (a, b, t) => a * (1 - t) + b * t;
+
+        const min_x = gsplatData.getProp('min_x', 'chunk');
+        const min_y = gsplatData.getProp('min_y', 'chunk');
+        const min_z = gsplatData.getProp('min_z', 'chunk');
+        const max_x = gsplatData.getProp('max_x', 'chunk');
+        const max_y = gsplatData.getProp('max_y', 'chunk');
+        const max_z = gsplatData.getProp('max_z', 'chunk');
+        const min_scale_x = gsplatData.getProp('min_scale_x', 'chunk');
+        const min_scale_y = gsplatData.getProp('min_scale_y', 'chunk');
+        const min_scale_z = gsplatData.getProp('min_scale_z', 'chunk');
+        const max_scale_x = gsplatData.getProp('max_scale_x', 'chunk');
+        const max_scale_y = gsplatData.getProp('max_scale_y', 'chunk');
+        const max_scale_z = gsplatData.getProp('max_scale_z', 'chunk');
+
+        const position = gsplatData.getProp('packed_position');
+        const rotation = gsplatData.getProp('packed_rotation');
+        const scale = gsplatData.getProp('packed_scale');
+        const color = gsplatData.getProp('packed_color');
+
+        this.read = (i) => {
```

Comment:
These functions will be allocated for each new instance, and V8 might be unable to optimize it as it depends on higher scope data. Which can lead to worse performance of that method due to deopt, especially taking into the account that it is called a lot of times. It is best to define it as a class member.

Reply:
Same goes to each method defined in constructor. It is best to have it as members of a class, instead of allocating new method for each instance. Also if it is possible to avoid calling a function for very high frequency methods (e.g. `lerp`), it is best to inline its code.

Reply:
I tried your suggestions, but don't see any difference in performance loading/decompressing/uploading a 6M splat scene. This is on chrome macos. (I would have been surprised if there was a difference tbh).

Reply:
With using inline lerp instead of a function, I get 3 times the performance in benchmarks, same goes for unpacking methods.
As we are dealing with millions of GS's, even small save can lead to significant loading time saves, so inlining these methods instead of using functions - does speed up it significantly.

Reply:
Hey @Maksims,

Yeah you're right. Speed really matters, and especially when loading compressed splats (the runtime case).

This code is really for the general case of accessing data. It's used when we decompress the data for use in editors like supersplat. It would be almost impossible to "micro-optimise" it to run fast because of the suboptimal structure.

So I've added a fast path for the functions we use at load time instead - the aabb calc and the splat center extraction. Load time processing has gone from 640ms -> 60ms.

I wouldn't have done this extra work if it wasn't for your nudging - so thank you so much!! <3<3